src/ColinHDev/CPlot/provider/DataProvider.php

```diff
@@ -1299,4 +1312,222 @@ public function savePlotRate(Plot $plot, PlotRate $plotRate) : Generator {
+                $playerData = yield $this->awaitPlayerDataByData(
+                    null,
+                    $XUID,
+                    $record["owner"]
+                );
+                if (!($playerData instanceof PlayerData)) {
+                    continue;
+                }
+
+                // load world
+                /** @var WorldSettings|false $world */
+                $world = yield $this->awaitWorld($record["level"]);
+                if($world === false)
+                    continue;
+
+                // load plot
+                /** @var Plot|null $plot */
+                $plot = yield $this->awaitPlot($record["level"], (int)$record["x"], (int)$record["z"]);
+                if($plot === null)
+                    continue;
+
+                // claim plot
+                $senderData = new PlotPlayer($playerData, PlotPlayer::STATE_OWNER);
+                $plot->addPlotPlayer($senderData);
+                yield from DataProvider::getInstance()->savePlotPlayer($plot, $senderData);
+
+				// load helpers
+				foreach($record["helpers"] as $playerName) {
+					// validate offline player data
+					$XUID = null;
+					$offlineData = Server::getInstance()->getOfflinePlayerData($playerName);
+					if($offlineData !== null) {
+						$XUID = $offlineData->getString("LastKnownXUID", "");
+
+						// register player data
+						yield from $this->updatePlayerData(
+							null, // doesn't matter what is input at this point. will overwrite on login
+							$XUID,
+							$playerName
+						);
+					}
+
+					$playerData = yield $this->awaitPlayerDataByData(
+						null,
+						$XUID,
+						$playerName
+					);
+					if (!($playerData instanceof PlayerData))
+						continue;
+
+					$senderData = new PlotPlayer($playerData, PlotPlayer::STATE_HELPER);
+					$plot->addPlotPlayer($senderData);
+					yield from DataProvider::getInstance()->savePlotPlayer($plot, $senderData);
+				}
+
+				// load denied with priority over helpers
+				foreach($record["denied"] as $playerName) {
+					// validate offline player data
+					$XUID = null;
+					$offlineData = Server::getInstance()->getOfflinePlayerData($playerName);
+					if($offlineData !== null) {
+						$XUID = $offlineData->getString("LastKnownXUID", "");
+
+						// register player data
+						yield from $this->updatePlayerData(
+							null, // doesn't matter what is input at this point. will overwrite on login
+							$XUID,
+							$playerName
+						);
+					}
+
+					$playerData = yield $this->awaitPlayerDataByData(
+						null,
+						$XUID,
+						$playerName
+					);
+					if (!($playerData instanceof PlayerData))
+						continue;
+
+					$senderData = new PlotPlayer($playerData, PlotPlayer::STATE_DENIED);
+					$plot->addPlotPlayer($senderData);
+					yield from DataProvider::getInstance()->savePlotPlayer($plot, $senderData);
+				}
+
+				//load common flags
+				$flag = FlagManager::getInstance()->getFlagByID("pvp");
```

Comment:
This can be shortened by using the [`Flags`](https://github.com/ColinHDev/CPlot/blob/development/src/ColinHDev/CPlot/plots/flags/Flags.php) class:
```php
$flag = Flags::PVP()->createInstance(...);
```